---
title: "SIB Days 2020 - Virtual Conference"
subtitle: "Breast Cancer. Aggregation"
author:
  - Patrick Roelli, Computational Biologist 2 - Computational Biology^[10x Genomics, patrick.roelli@10xgenomics.com ]
  - Stefania Giacomello, Computational Biologist 2 - Computational Biology^[10x Genomics, stephen.williams@10xgenomics.com]
  - Stephen Williams, Senior Scientist - Computational Biology^[10x Genomics, stephen.williams@10xgenomics.com]
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_notebook:
    code_folding: none
    theme: journal
    toc: yes
    toc_depth: 3
    toc_float: yes
---

# **Introduction**

Welcome to the **SIB Days 2020 - virtual conference** Spatial Transcriptomics workshop by 10x genomics!

[![10x Home](https://github.com/stephenwilliams22/SIB_2020_Workshop/raw/master/images/10x%20homepage.png)](https://www.10xgenomics.com/)


The purpose of this tutorial will be to walk users through some of the steps necessary to explore data produced by the 10x Genomics Visium Spatail Gene Expression Solution and the [Spaceranger pipeline](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger). We will investigate the datasets whith are all freely available from [10x Genomics](https://support.10xgenomics.com/spatial-gene-expression/datasets).

[Seurat Tutorial](https://satijalab.org/seurat/v3.1/spatial_vignette.html)

**Things to know about this workshop**

1. All files that will be used can be found at: `/mnt/libs/shared_data/`
2. Getting started with R and Visium data outside of Seurat at: https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/rkit
3. Reference genome for all samples is GRCh38/mm10
4. All 10x software including [Spaceranger](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/what-is-space-ranger), and [Loupe Browser](https://support.10xgenomics.com/spatial-gene-expression/software/visualization/latest/what-is-loupe-browser), can be downloaded from the [10x Support Site](https://support.10xgenomics.com/) 


# **Exploring Visium Data with Seurat**
## Load our packages
```{r Libraries, echo=TRUE, message=FALSE, warning=FALSE}
library(Seurat)
library(ggplot2)
library(patchwork)
library(dplyr)
library(RColorBrewer)
```

## Loading data in a Seurat object

Real Dataset for the tutorial
```{r eval=FALSE}
breast_cancer_1 <- Load10X_Spatial(data.dir = "/mnt/libs/shared_data/human_breast_cancer_1/outs/",
                filename = "V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5")

breast_cancer_2 <- Load10X_Spatial(data.dir = "/mnt/libs/shared_data/human_breast_cancer_2/outs/",
                filename = "V1_Breast_Cancer_Block_A_Section_2_filtered_feature_bc_matrix.h5")
```


Same data just internal to 10x
```{r}
breast_cancer_1 <- Load10X_Spatial(data.dir = "/mnt/analysis/marsoc/pipestances/HWHTFDSXX/SPATIAL_RNA_COUNTER_PD/163086/HEAD/outs/", slice = "slice1")
breast_cancer_2 <- Load10X_Spatial(data.dir = "/mnt/analysis/marsoc/pipestances/HWHTFDSXX/SPATIAL_RNA_COUNTER_PD/163087/HEAD/outs/", slice = "slice2")
```

Let's merge the data together
```{r}
breast_cancer <- merge(breast_cancer_1, breast_cancer_2)
```


There are a bunch of datasets hoted by the Satija lab in the [Seurat Data Package](https://github.com/satijalab/seurat-data).

# Results
## QC
Let's have a look at some basic QC information. Keep in mind that most seurat plots are ggplot object and can be manipulated as such.

Counts = UMI
Features = Genes
```{r, fig.width=10,fig.height=10, warning=FALSE}
plot1 <- VlnPlot(breast_cancer, features = "nCount_Spatial", pt.size = 0.1) + 
  ggtitle("UMI") +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(), 
        legend.position = "right") +
  NoLegend()

plot2 <- VlnPlot(breast_cancer, features = "nFeature_Spatial", pt.size = 0.1) + 
  ggtitle("Genes") +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(), 
        legend.position = "right") +
  NoLegend()

plot3 <- SpatialFeaturePlot(breast_cancer, features = "nCount_Spatial") + 
  theme(legend.position = "right")

plot4 <- SpatialFeaturePlot(breast_cancer, features = "nFeature_Spatial") +
  theme(legend.position = "right")

plot1 + plot2 + plot3 + plot_spacer() + plot4 + plot_layout(nrow = 3, ncol = 2)
```

## Normilization

Spaceranger does normiliaztion for clustering and DE but does not return that normalized matrix

Pre-normalization 
Raw UMI counts
```{r, fig.width=10, warning=FALSE}
SpatialFeaturePlot(breast_cancer, features = c("ERBB2", "CD8A"))
```
SE transform

+  This will take ~3-4 min. 

Don't worry about `reachediteration limit` warnings. See https://github.com/ChristophH/sctransform/issues/25 for discussion

Default assay will now be set to SCT
```{r, warning=FALSE}
breast_cancer <- SCTransform(breast_cancer, assay = "Spatial", verbose = TRUE)
```

```{r, fig.width=10, warning=FALSE}
SpatialFeaturePlot(breast_cancer, features = c("ERBB2", "CD8A"))
```


From Seurat: 

The default parameters in Seurat emphasize the visualization of molecular data. However, you can also adjust the size of the spots (and their transparency) to improve the visualization of the histology image, by changing the following parameters:

    pt.size.factor- This will scale the size of the spots. Default is 1.6
    alpha - minimum and maximum transparency. Default is c(1, 1).
    Try setting to alpha c(0.1, 1), to downweight the transparency of points with lower expression



```{r, fig.width=10, warning=FALSE}
p1 <- SpatialFeaturePlot(breast_cancer, features = "TTR", pt.size.factor = 1)+ 
  theme(legend.position = "right") +
  ggtitle("Actual Spot Size")
p2 <- SpatialFeaturePlot(breast_cancer, features = "TTR")+ 
  theme(legend.position = "right") +
  ggtitle("Scaled Spot Size")
p1 + p2 + plot_annotation(
  title = 'Actual Spot Size (left), Scaled Spot Size (right)'
)
```

Dimensionality reduction, clustering, and visualization

We can then proceed to run dimensionality reduction and clustering on the RNA expression data, using the same workflow as we use for scRNA-seq analysis.

Some of these processes can be parallized

```{r}
library(future)
# check the current active plan
plan()
```

```{r}
# change the current plan to access parallelization
plan("multiprocess", workers = 4)
plan()
```


The defalut UMAP calculation is performed with the [R-based UWOT](https://cran.r-project.org/web/packages/uwot/index.html) library However, you can run UMAP in python via retuculate library and `umap-learn`. We have found that for smaller datasets (<= 10k cells/spots) UWOT is great. For much larger datasets (100k + cells/spots) `umap-learn` can be a faster option. 
```{r}
breast_cancer <- RunPCA(breast_cancer, assay = "SCT", verbose = FALSE)
breast_cancer <- FindNeighbors(breast_cancer, reduction = "pca", dims = 1:30)
breast_cancer <- FindClusters(breast_cancer, verbose = FALSE)
breast_cancer <- RunUMAP(breast_cancer, reduction = "pca", dims = 1:30)
```

Now let's have a look at the clustering

First let's extract the index of the barcodes and add it to the metadata
```{r, fig.width=12}
substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}

breast_cancer@meta.data$merged.ident <- paste("Slice_", substrRight(rownames(breast_cancer@meta.data), 1), sep = "")
```

Let's have a look. It looks like there isn't any large difference between the two slices that might make us want to do some sort of normilization for batch effects. There is a really nice paper on integration from a single cell perspective that was published on bioRxiv recently. [MD Luecken et. al](https://www.biorxiv.org/content/10.1101/2020.05.22.111161v1)

```{r, fig.width=12}
DimPlot(breast_cancer, reduction = "umap", label = FALSE, group.by = c("ident", "merged.ident")) +
  labs(color = "Cluster")
```


I don't really like these colors so let's change them
```{r, fig.width=10,fig.height=10, warning=FALSE}
p1 <- DimPlot(breast_cancer, reduction = "umap", label = TRUE) +
  labs(color = "Cluster")
p2 <- SpatialDimPlot(breast_cancer, label = TRUE, label.size = 3) +
  labs(fill = "Cluster")

p1 + p2 + plot_annotation(
  title = 'Clustering in UMAP and Tissue Space',
  caption = 'Processed by Spaceranger 1.1\nNormilization and Clustering by Seurat'
) + plot_layout(nrow = 2)
```


```{r}
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
```

```{r, fig.width=10, fig.height=10, warning=FALSE}
p1 <- DimPlot(breast_cancer, reduction = "umap", label = TRUE) +
  labs(color = "Cluster") + 
  scale_color_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                               "#a65628", "#999999", "black", "pink", "purple", "brown",
                               "grey", "yellow", "green"))

p2 <- SpatialDimPlot(breast_cancer, label = TRUE, label.size = 3) +
  labs(fill = "Cluster")

p1 + p2 + plot_annotation(
  title = 'Clustering in UMAP and Tissue Space',
  caption = 'Processed by Spaceranger 1.1\nNormilization and Clustering by Seurat'
) + plot_layout(nrow = 2)
```

Interactivity not working for me on firefox
```{r}
LinkedDimPlot(breast_cancer)
```

## Spatially variable features

First we'll idetify differentially expressed genes. 

Parallelization helps here too let's make sure our plan is still intact

```{r}
plan()
```
`- call: plan("multiprocess", workers = 4)` indicates that it is

Looks like we have some very DE genes for clusters 4 and 11


clarify what ident.1 = 4, ident.2 = 6 are for
```{r, fig.width=10, warning=FALSE}
de_markers <- FindMarkers(breast_cancer, ident.1 = 4, ident.2 = 6)
```


```{r, fig.width=10, fig.height=10, warning=FALSE}
SpatialFeaturePlot(object = breast_cancer, features = rownames(de_markers)[1:3], alpha = c(0.1, 1), ncol = 3)
```

what are the top variable features?
```{r}
VariableFeatures(breast_cancer)[1:10]
```

what are the top de genes?
```{r}
rownames(de_markers)[1:10]
```

## Spatially Variable Genes

So what about spatial enrichment? 

Some methods
1. [Trendsceek](https://www.nature.com/articles/nmeth.4634)
2. [Splotch](https://www.biorxiv.org/content/10.1101/757096v1)
3. [SPARK](https://www.nature.com/articles/s41592-019-0701-7)
4. [SpatialDE](https://www.nature.com/articles/nmeth.4636)
  + We have found this implimenton not to be very effective. It's also not under active development


Using the top 100 variable genes find spatially enriched ones. Note that in the Seurat Spatial Tutorial they use 1000 genes. You can also use all genes but that will take a long time. Using a calucation of [Morans I](https://en.wikipedia.org/wiki/Moran%27s_I) can sometimes be a faster approach, especially if you are using parallization.

This falls apart for merged data. 

```{r}
breast_cancer <- FindSpatiallyVariableFeatures(breast_cancer, 
                                               assay = "SCT", 
                                               slot = "scale.data", 
                                               features = VariableFeatures(breast_cancer)[1:100],
                                               selection.method = "markvariogram", verbose = TRUE)
```

Have a look at the spatially variable genes calculated by `markvariogram` ordered from most variable to least variable
```{r}
SpatiallyVariableFeatures(breast_cancer, selection.method = "markvariogram", decreasing = TRUE)
```

```{r, fig.width=10, fig.height=8}
top.features_trendseq <- head(SpatiallyVariableFeatures(breast_cancer, selection.method = "markvariogram"), 6)
SpatialFeaturePlot(breast_cancer, features = top.features_trendseq, ncol = 3, alpha = c(0.1, 1))
```

Moran's I implamentation. For other spatial data types the x.cuts and y.cuts determins the grid that is laied over the tissue in the capture area. Here we'll remove those
```{r}
breast_cancer <- FindSpatiallyVariableFeatures(breast_cancer, assay = "SCT", slot = "scale.data", features = VariableFeatures(breast_cancer)[1:100], 
    selection.method = "moransi")
```


Have a look at the spatially variable genes calculated by `moransi` ordered from most variable to least variable

```{r}
SpatiallyVariableFeatures(breast_cancer, selection.method = "moransi", decreasing = TRUE)
```

```{r, fig.width=10, fig.height=8}
top.features_moransi <- head(SpatiallyVariableFeatures(breast_cancer, selection.method = "moransi"), 8)
SpatialFeaturePlot(breast_cancer, features = top.features_moransi, ncol = 4, alpha = c(0.1, 1))
```

We can see that the results are slightly different. So let's take a look at why.

```{r}
spatially_variable_genes <- breast_cancer@assays$SCT@meta.features %>%
  tidyr::drop_na()

spatially_variable_genes
```
You can see the two methods show 
```{r}
mm_cor <- cor.test(spatially_variable_genes$moransi.spatially.variable.rank, spatially_variable_genes$markvariogram.spatially.variable.rank)
ggplot(spatially_variable_genes, aes(x=moransi.spatially.variable.rank,y=markvariogram.spatially.variable.rank))+
  geom_point()+
  geom_smooth()+
  xlab("Morans I Rank")+
  ylab("Markvariogram Rank")+
  annotate("text", x = 25, y = 75, label = paste("Pearson's Correlation\n", round(mm_cor$estimate[1], digits = 2), sep = ""))+
  theme_bw()
```



```{r}
images_cl <- list()

for (i in 1:length(lenas)) {
  images_cl[[i]] <- readbitmap::read.bitmap(paste(get_spatial_path(lenas[i]), "/outs/spatial/tissue_lowres_image.png", sep = ""))
}
```


```{r}
height <- list()

for (i in 1:length(lenas)) {
 height[[i]] <-  data.frame(height = nrow(images_cl[[i]]))
}

height <- bind_rows(height)

width <- list()

for (i in 1:length(lenas)) {
 width[[i]] <- data.frame(width = ncol(images_cl[[i]]))
}

width <- bind_rows(width)
```


###Color from pipeline
```{r}
grobs <- list()

for (i in 1:length(lenas)) {
  grobs[[i]] <- grid::rasterGrob(images_cl[[i]], width=unit(1,"npc"), height=unit(1,"npc"))
}

images_tibble <- tibble(lena=lenas, grob=grobs)
images_tibble$lena <- factor(images_tibble$lena)
images_tibble$height <- height$height
images_tibble$width <- width$width
images_tibble
```

```{r}
scales <- list()

for (i in 1:length(lenas)) {
 path_scales <- paste(svenLib::get_spatial_path(lenas[i]), "/outs/spatial/scalefactors_json.json", sep = "")
 scales[[i]] <- rjson::fromJSON(file = path_scales)
}
```

### Clusters and tsne
```{r}
clusters <- list()
for (i in 1:length(lenas)) {
  clusters[[i]] <- read.csv(paste(svenLib::get_spatial_path(lenas[i]),"/outs/analysis_csv/clustering/graphclust/clusters.csv", sep = ""))
}


tsne <- list()

for (i in 1:length(lenas)) {
   tsne[[i]] <- read.csv(paste(svenLib::get_spatial_path(lenas[i]),"/outs/analysis_csv/tsne/2_components/projection.csv",sep = ""), header = T)
}

umap <- list()

for (i in 1:length(lenas)) {
   umap[[i]] <- read.csv(paste(svenLib::get_spatial_path(lenas[i]),"/outs/analysis_csv/umap/2_components/projection.csv",sep = ""), header = T)
}

umap[[1]]
```


### Combine clusters, tsne, and tissue info for easy plotting
```{r}
bcs <- list()

for (i in 1:length(lenas)) {
   if (file.exists(paste(get_spatial_path(lenas[i]),"/outs/spatial/tissue_positions_list.txt", sep = ""))) {
  bcs[[i]] <- read.csv(paste(get_spatial_path(lenas[i]), "/outs/spatial/tissue_positions_list.txt", sep = ""),
                 col.names=c("barcode","tissue","row","col","imagerow","imagecol"), header = F)
        } else {
  bcs[[i]] <- read.csv(paste(get_spatial_path(lenas[i]), "/outs/spatial/tissue_positions_list.csv", sep = ""),
                 col.names=c( "barcode","tissue","row","col","imagerow","imagecol"), header = F)
        }
   bcs[[i]]$imagerow_scaled <- bcs[[i]]$imagerow * scales[[i]]$tissue_lowres_scalef    # scale tissue coordinates for lowres image
   bcs[[i]]$imagecol_scaled <- bcs[[i]]$imagecol * scales[[i]]$tissue_lowres_scalef
   bcs[[i]]$imagerow_scaled_round <- round(bcs[[i]]$imagerow * scales[[i]]$tissue_lowres_scalef) # Rounded scales
   bcs[[i]]$imagecol_scaled_round <- round(bcs[[i]]$imagecol * scales[[i]]$tissue_lowres_scalef)
   bcs[[i]]$tissue <- as.factor(bcs[[i]]$tissue)
   bcs[[i]] <- merge(bcs[[i]], clusters[[i]], by.x = "barcode", by.y = "Barcode", all = TRUE)
   bcs[[i]] <- merge(bcs[[i]], tsne[[i]], by.x = "barcode", by.y = "Barcode", all = TRUE)
   bcs[[i]] <- merge(bcs[[i]], umap[[i]], by.x = "barcode", by.y = "Barcode", all = TRUE)
   bcs[[i]]$height <- height$height[i]
   bcs[[i]]$width <- width$width[i]
}

names(bcs) <- lenas
```


```{r}
read_matrix <- function(sid) {
  Matrix::t(Seurat::Read10X_h5(paste(get_spatial_path(sid = sid), "/outs/raw_feature_bc_matrix.h5", sep = "")))
}

matrix <- map(.x = lenas, .f = read_matrix)
names(matrix) <- lenas
matrix[[1]]
```


```{r}
umi_sum <- list() 

for (i in 1:length(lenas)) {
  umi_sum[[i]] <- data.frame(barcode =  row.names(matrix[[i]]),
                             sum_umi = Matrix::rowSums(matrix[[i]]))
  
}
names(umi_sum) <- lenas

umi_sum <- bind_rows(umi_sum, .id = "lena")
umi_sum


gene_sum <- list() 

for (i in 1:length(lenas)) {
  gene_sum[[i]] <- data.frame(barcode =  row.names(matrix[[i]]),
                             sum_gene = Matrix::rowSums(matrix[[i]] != 0))
  
}
names(gene_sum) <- lenas

gene_sum <- bind_rows(gene_sum, .id = "lena")
gene_sum

# If you need to look at the correlation of gene expression between samples
gene_umi_sum <- list() 

for (i in 1:length(lenas)) {
  gene_umi_sum[[i]] <- data.frame(gene =  colnames(matrix[[i]]),
                             gene_umi_sum = Matrix::colSums(matrix[[i]]))
  
}
names(gene_umi_sum) <- lenas

gene_umi_sum <- bind_rows(gene_umi_sum, .id = "lena")
gene_umi_sum
```

```{r}
bcs_merge <- bind_rows(bcs, .id = "lena")
bcs_merge <- merge(bcs_merge,umi_sum, by = c("barcode", "lena"))
bcs_merge <- merge(bcs_merge,gene_sum, by = c("barcode", "lena"))
```

Define our color palette for plotting
```{r}
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
```

```{r, fig.width = 18, fig.height = 9}
plots <- list()

for (i in 1:length(lenas)) {

plots[[i]] <- bcs_merge %>% 
  filter(lena ==lenas[i]) %>% 
  filter(tissue =="1") %>% 
      ggplot(aes(x=imagecol_scaled,y=imagerow_scaled,fill=sum_umi)) +
                geom_spatial(data=images_tibble[i,], aes(grob=grob), x=0.5, y=0.5)+
                geom_point(shape = 21, colour = "black", size = 2, stroke = 0.1)+
                coord_cartesian(expand=FALSE)+
                scale_fill_gradientn(colours = myPalette(100))+
                #facet_wrap(~lena)+
                xlim(0,max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(width)))+
                ylim(max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(height)),0)+
                xlab("") +
                ylab("") +
                ggtitle(paste(lenas[i],": ", sample_type[i], sep = ""))+
                labs(fill = "UMI")+
                theme_set(theme_bw(base_size = 10))+
                theme(panel.grid.major = element_blank(), 
                        panel.grid.minor = element_blank(),
                        panel.background = element_blank(), 
                        axis.line = element_line(colour = "black"),
                        axis.text = element_blank())
}

patchwork::wrap_plots(plots)
```

```{r, fig.width = 18, fig.height = 9}
plots <- list()

for (i in 1:length(lenas)) {

plots[[i]] <- bcs_merge %>% 
  filter(lena ==lenas[i]) %>% 
  filter(tissue =="1") %>% 
      ggplot(aes(x=imagecol_scaled,y=imagerow_scaled,fill=sum_gene)) +
                geom_spatial(data=images_tibble[i,], aes(grob=grob), x=0.5, y=0.5)+
                geom_point(shape = 21, colour = "black", size = 2, stroke = 0.1)+
                coord_cartesian(expand=FALSE)+
                scale_fill_gradientn(colours = myPalette(100))+
                #facet_wrap(~lena)+
                xlim(0,max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(width)))+
                ylim(max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(height)),0)+
                xlab("") +
                ylab("") +
                ggtitle(paste(lenas[i],": ", sample_type[i], sep = ""))+
                labs(fill = "Genes")+
                theme_set(theme_bw(base_size = 10))+
                theme(panel.grid.major = element_blank(), 
                        panel.grid.minor = element_blank(),
                        panel.background = element_blank(), 
                        axis.line = element_line(colour = "black"),
                        axis.text = element_blank())
}

patchwork::wrap_plots(plots)
```

```{r, fig.width = 18, fig.height = 9}
plots <- list()

for (i in 1:length(lenas)) {

plots[[i]] <- bcs_merge %>% 
  filter(lena ==lenas[i]) %>%
  add_column(GAPDH = matrix[[i]][,"GAPDH"]) %>% 
      ggplot(aes(x=imagecol_scaled,y=imagerow_scaled,fill=GAPDH)) +
                geom_spatial(data=images_tibble[i,], aes(grob=grob), x=0.5, y=0.5)+
                geom_point(shape = 21, colour = "black", size = 2, stroke = 0.1)+
                coord_cartesian(expand=FALSE)+
                scale_fill_gradientn(colours = myPalette(100))+
                #facet_wrap(~lena)+
                xlim(0,max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(width)))+
                ylim(max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(height)),0)+
                xlab("") +
                ylab("") +
                ggtitle(paste(lenas[i],": ", sample_type[i], sep = ""))+
                labs(fill = "GAPDH UMI")+
                theme_set(theme_bw(base_size = 10))+
                theme(panel.grid.major = element_blank(), 
                        panel.grid.minor = element_blank(),
                        panel.background = element_blank(), 
                        axis.line = element_line(colour = "black"),
                        axis.text = element_blank())
}

patchwork::wrap_plots(plots)
```


```{r, fig.width = 18, fig.height = 9}
plots <- list()

for (i in 1:length(lenas)) {

plots[[i]] <- bcs_merge %>% 
  filter(lena ==lenas[i]) %>%
  filter(tissue == "1") %>% 
  na.omit() %>% 
      ggplot(aes(x=imagecol_scaled,y=imagerow_scaled,fill=factor(Cluster))) +
                geom_spatial(data=images_tibble[i,], aes(grob=grob), x=0.5, y=0.5)+
                geom_point(shape = 21, colour = "black", size = 2, stroke = 0.1)+
                coord_cartesian(expand=FALSE)+
                scale_fill_manual(values = c("#b2df8a","#e41a1c","#377eb8","#4daf4a","#ff7f00","gold", 
                                             "#a65628", "#999999", "black", "white", "purple", "brown"))+
                xlim(0,max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(width)))+
                ylim(max(bcs_merge %>% 
                            filter(lena ==lenas[i]) %>% 
                            dplyr::select(height)),0)+
                xlab("") +
                ylab("") +
                ggtitle(paste(lenas[i],": ", sample_type[i], sep = ""))+
                labs(fill = "Cluster")+
                guides(fill = guide_legend(override.aes = list(size=3)))+
                theme_set(theme_bw(base_size = 10))+
                theme(panel.grid.major = element_blank(), 
                        panel.grid.minor = element_blank(),
                        panel.background = element_blank(), 
                        axis.line = element_line(colour = "black"),
                        axis.text = element_blank())
  
}
patchwork::wrap_plots(plots)
```


